cpp/src/arrow/util/bit_block_counter.h

```diff
@@ -547,24 +548,128 @@ static void VisitTwoBitBlocksVoid(const uint8_t* left_bitmap, int64_t left_offse
     BitBlockCount block = bit_counter.NextAndWord();
     if (block.AllSet()) {
       for (int64_t i = 0; i < block.length; ++i, ++position) {
-        visit_not_null(position);
+        InvokeWithRequiredArgs(visit_not_null, position);
       }
     } else if (block.NoneSet()) {
       for (int64_t i = 0; i < block.length; ++i, ++position) {
-        visit_null();
+        InvokeWithRequiredArgs(visit_null, position);
       }
     } else {
       for (int64_t i = 0; i < block.length; ++i, ++position) {
         if (bit_util::GetBit(left_bitmap, left_offset + position) &&
             bit_util::GetBit(right_bitmap, right_offset + position)) {
-          visit_not_null(position);
+          InvokeWithRequiredArgs(visit_not_null, position);
         } else {
-          visit_null();
+          InvokeWithRequiredArgs(visit_null, position);
         }
       }
     }
   }
 }
 
-}  // namespace internal
-}  // namespace arrow
+template <typename VisitBothNotNull, typename VisitLeftNull, typename VisitRightNull,
+          typename VisitBothNull>
+static Status VisitTwoBitBlocksAllCases(const uint8_t* left_bitmap, int64_t left_offset,
+                                        const uint8_t* right_bitmap, int64_t right_offset,
+                                        int64_t length,
+                                        VisitBothNotNull&& visit_both_not_null,
+                                        VisitLeftNull&& visit_left_null,
+                                        VisitRightNull&& visit_right_null,
+                                        VisitBothNull&& visit_both_null) {
```

Comment:
The current `visit_valid` must have 1 arg and `visit_null` must have no args. However, I want to pass `VisitRightNull` and `VisitLeftNull` (both having 0 args) as `visit_valid`, so I need to use InvokeWithRequiredArgs for all the lambdas passed.